eval/topdown.go

```diff
@@ -159,21 +194,69 @@ var builtinFunctions = map[opalog.Var]builtinFunction{
 // dereferenceVar is used to lookup the variable binding and convert the value to
 // a native Go type.
 func dereferenceVar(v opalog.Var, ctx *TopDownContext) (interface{}, error) {
-	binding, ok := ctx.Bindings[v]
-	if !ok {
+	binding := ctx.Bindings.Get(v)
+	if binding == nil {
 		return nil, fmt.Errorf("unbound variable: %v", v)
 	}
-
 	return valueToInterface(binding, ctx)
 }
 
+func evalContext(ctx *TopDownContext, iter TopDownIterator) error {
+
+	if ctx.Index >= len(ctx.Query) {
+		ctx.traceFinish()
+		return iter(ctx)
+	}
+
+	ctx.traceEval()
+
+	if ctx.Current().Negated {
+		return evalContextNegated(ctx, iter)
+	}
+
+	return evalTerms(ctx, func(ctx *TopDownContext) error {
+		return evalExpr(ctx, func(ctx *TopDownContext) error {
+			ctx = ctx.Step()
+			return evalContext(ctx, iter)
+		})
+	})
+}
+
+func evalContextNegated(ctx *TopDownContext, iter TopDownIterator) error {
+
+	negation := *ctx
+	negation.Query = opalog.Body([]*opalog.Expr{ctx.Current().Complement()})
+	negation.Index = 0
+	negation.Previous = ctx
+
+	isDefined := false
+
+	err := evalContext(&negation, func(*TopDownContext) error {
+		isDefined = true
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !isDefined {
+		return evalContext(ctx.Step(), iter)
+	}
+
+	return nil
+}
+
 func evalEq(ctx *TopDownContext, expr *opalog.Expr, iter TopDownIterator) error {
 
 	operands := expr.Terms.([]*opalog.Term)
 	a := operands[1].Value
 	b := operands[2].Value
 
-	return evalEqUnify(ctx, a, b, iter)
+	return evalEqUnify(ctx, a, b, func(ctx *TopDownContext) error {
+		ctx.traceSuccess(expr)
```

Comment:
Seems awkward to create anonymous functions everywhere just for tracing.  Any way to fold the tracing into the code more natively?  (Every anon function I see takes a while for me to grok.  It'll be worse for people who don't know the alg.)